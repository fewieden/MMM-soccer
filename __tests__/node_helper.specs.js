const {generateResponse} = require('../__mocks__/mockResponse');

describe('node_helper', () => {
    let helper;

    beforeEach(() => {
        helper = require('../node_helper');

        helper.setName('MMM-soccer');

        fetchMock.mockResponseOnce(JSON.stringify(generateResponse()), {status: 200});
    });

    test('start prints module name', () => {
        helper.start();

        expect(console.log).toHaveBeenCalledWith('Starting module helper: MMM-soccer');
    });

    test('triggers data fetch without api_key if notification is GET_DATA', () => {
        helper.socketNotificationReceived('GET_DATA', {league: 'BL1'});

        expect(fetchMock).toHaveBeenNthCalledWith(1, 'http://api.football-data.org/v4/competitions/BL1/standings', {});
    });

    test('triggers data fetch with api key if notification is GET_DATA', () => {
        const apiKey = 'TEST_API_KEY';
        helper.socketNotificationReceived('GET_DATA', {league: 'BL1', api_key: apiKey});

        expect(fetchMock).toHaveBeenNthCalledWith(1, 'http://api.football-data.org/v4/competitions/BL1/standings', {
            headers: {
                'X-Auth-Token': apiKey
            }
        });
    });

    test('data fetching sends response via socket to module on success', async() => {
        helper.socketNotificationReceived('GET_DATA', {league: 'BL1'});

        await waitForAsync();

        expect(helper.sendSocketNotification.mock.calls[0][0]).toBe('DATA');
        expect(helper.sendSocketNotification.mock.calls[0][1]).toMatchSnapshot();
    });

    test('data fetching logs error on failure', async() => {
        fetchMock.resetMocks();
        fetchMock.mockResponseOnce(JSON.stringify({message: 'Forbidden'}), {status: 403});

        helper.socketNotificationReceived('GET_DATA', {league: 'BL1'});

        await waitForAsync();

        expect(console.error).toHaveBeenCalledWith('Getting league table: 403 Forbidden');
        expect(helper.sendSocketNotification).not.toHaveBeenCalled();
    });

    test('does NOT trigger data fetch if notification is different than GET_DATA', () => {
        helper.socketNotificationReceived('OTHER NOTIFICATION');

        expect(fetchMock).not.toHaveBeenCalled();
    });
});
