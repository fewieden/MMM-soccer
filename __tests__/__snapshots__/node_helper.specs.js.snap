// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`node_helper data fetching sends response via socket to module on success 1`] = `
{
  "competition": {
    "area": {
      "id": 2088,
      "name": "Germany",
    },
    "code": "BL1",
    "id": 2002,
    "lastUpdated": "2021-01-03T01:25:08Z",
    "name": "Bundesliga",
    "plan": "TIER_ONE",
  },
  "season": {
    "currentMatchday": 14,
    "endDate": "2021-05-15",
    "id": 599,
    "startDate": "2020-09-18",
  },
  "standings": [
    {
      "stage": "REGULAR_SEASON",
      "table": [
        {
          "draw": 4,
          "form": "W,D,W,W,D",
          "goalDifference": 16,
          "goalsAgainst": 9,
          "goalsFor": 25,
          "lost": 1,
          "playedGames": 14,
          "points": 31,
          "position": 1,
          "team": {
            "crestUrl": "https://crests.football-data.org/721.svg",
            "id": 721,
            "name": "RB Leipzig",
          },
          "won": 9,
        },
        {
          "draw": 3,
          "form": "W,W,D,D,W",
          "goalDifference": 20,
          "goalsAgainst": 19,
          "goalsFor": 39,
          "lost": 1,
          "playedGames": 13,
          "points": 30,
          "position": 2,
          "team": {
            "crestUrl": "https://crests.football-data.org/5.svg",
            "id": 5,
            "name": "FC Bayern MÃ¼nchen",
          },
          "won": 9,
        },
        {
          "draw": 4,
          "form": "L,L,W,W,W",
          "goalDifference": 15,
          "goalsAgainst": 14,
          "goalsFor": 29,
          "lost": 2,
          "playedGames": 14,
          "points": 28,
          "position": 3,
          "team": {
            "crestUrl": "https://crests.football-data.org/3.svg",
            "id": 3,
            "name": "Bayer 04 Leverkusen",
          },
          "won": 8,
        },
      ],
      "type": "TOTAL",
    },
  ],
}
`;
